src/design-pages/passwordless_gdm.rst

```diff
@@ -0,0 +1,461 @@
+    }
+
+The field meaning is as follows:
+
+* name: the mechanism name that will be shown in the login screen.
+* role: the authentication mechanism. It can be password, EIdP, smartcard or
+  passkey.
+* selectable: whether the authentication mechanism should be shown up in the
+  login screen as an explicitly selectable option. By default it is set to
+  ``true``.
+
+GDM answers with the result for the previous JSON message processing:
+
+..  code-block:: json
+
+    {
+      "auth-selection": {
+        "status": "$status",
+        "mech": {
+          "data1": "$data1",
+          "data2": "$data2"
+        }
+      }
+    }
+
+For clarification, there are two different types of user interaction messages:
+`prompt` and `instruction`. The first one asks the user to enter something on
+the screen, e.g. the PIN. The second type prompts the user to follow
+instructions, e.g. to enter the PIN into the reader.
+
+Examples for the two messages are defined in the following lines.
+
+Password
+++++++++
+SSSD provides the available authentication mechanisms:
+
+..  code-block:: json
+
+    {
+      "auth-selection": {
+        "mechanisms": {
+          "password": {
+            "name": "Password",
+            "role": "password",
+            "selectable": true,
+            "prompt": "Password:"
+          }
+        },
+        "priority": [
+          "password"
+        ]
+      }
+    }
+
+GDM replies with the selection:
+
+..  code-block:: json
+
+    {
+      "auth-selection": {
+        "status": "Ok",
+        "password": {
+          "password": "ThePassword"
+        }
+      }
+    }
+
+EIdP
+++++
+SSSD provides the available authentication mechanisms:
+
+..  code-block:: json
+
+    {
+      "auth-selection": {
+        "mechanisms": {
+          "eidp": {
+            "name": "Web Login",
+            "role": "eidp",
+            "selectable": true,
+            "init_prompt": "Login",
+            "link_prompt": "Log in online with another device",
+            "uri": "https://short.url.com/1234",
+            "code": "1234",
+            "timeout": 300
+          }
+        },
+        "priority": [
+          "eidp"
+        ]
+      }
+    }
+
+GDM replies with the selection:
+
+..  code-block:: json
+
+    {
+      "auth-selection": {
+        "status": "Ok",
+        "eidp": {}
+      }
+    }
+
+Smart card
+++++++++++
+SSSD provides the available authentication mechanisms:
+
+..  code-block:: json
+
+    {
+      "auth-selection": {
+        "mechanisms": {
+          "smartcard:1": {
+            "name": "smartcard ID1",
+            "role": "smartcard",
+            "selectable": true,
+            "init_prompt": "Insert smartcard",
+            "pin_prompt": "PIN"
+          },
+          "smartcard:2": {
+            "name": "smartcard ID2",
+            "role": "smartcard",
+            "selectable": true,
+            "init_prompt": "Insert smartcard",
+            "pin_instruction": "Enter PIN in reader"
+          }
```

Comment:
Hi,

I think an "instruction prompt" is missing here, at least if there are multiple matching certificates for the user. It is what is called just `prompt` in SSSD's `pack_cert_data()` and is current a combination of the certificate label and the subject DN of the certificate. To be able to select the right certificate this string should be displayed together with the other details of the given method.

bye,
Sumit

Reply:
That's taken care of in the [name](https://github.com/SSSD/sssd.io/pull/79/files#diff-3682413448fe17192d0456526586fa27b75df6ad1dea69372800f974e4203c02R266) attribute. I can change the example to make it more clear, but for that I'll need your help as so far I didn't find any good example.

Reply:
Hi,
you can find examples in the PAM responder unit tests (https://github.com/SSSD/sssd/blob/master/src/tests/cmocka/test_pam_srv.c#L87), so e.g.

    "SSSD test cert 0001\nCN=SSSD test cert 0001,OU=SSSD test,O=SSSD"

Please note the `\n`, so on the console it is displayed on two lines, first line the Smartcard label and the second line the subject DN of the certificate. Currently this is hard-coded but it might be worth to make the string configurable. So GDM should not expect a fixed number of lines.

bye,
Sumit

Reply:
Actually, this part of the code is already implemented and I am using the [label](https://github.com/SSSD/sssd/blob/master/src/sss_client/pam_sss.c#L135) field of the cert_auth_info structure. Does it contain the label and the subject DN, or just the label?

Reply:
Hi,

better would be `prompt_str` because `label` would be the same for all certificates from the same Smartcard. `prompt_str` is used in the [current GDM related code](https://github.com/SSSD/sssd/blob/master/src/sss_client/pam_sss.c#L2060) and for[ text based prompting](https://github.com/SSSD/sssd/blob/master/src/sss_client/pam_sss.c#L2152).

bye,
Sumit

Reply:
So I'll do that for the implementation. Does this part of the design seem clear and OK to you?

Reply:
Hi,

yes, I'm fine, it just might be worth to make the GDM developers aware that the `name` field might have multiple lines.

bye,
Sumit

Reply:
@halfline just so you're aware, smartcard's `name` field might have multiple lines.

Sumit, I updated the design to include this information, so I'm resolving the conversation.